### PR TITLE
[WFMP-154] Integrate Channels to provision WildFly

### PR DESCRIPTION
**Pull request by jmesnil**
JIRA: https://issues.redhat.com/browse/WFMP-154

### CODE REVIEW

**Comment by jmesnil** [comment]
the JDK8 Jobs fails as the integrated `wildfly-channel` targets Java 11: https://github.com/wildfly-extras/wildfly-channel/blob/2e798760d37da1290eaa8a7d22526cdf1db5fb1a/pom.xml#L26

File: core/src/main/java/org/wildfly/plugin/core/GalleonUtils.java [hunk 1]
```diff
@@ -149,7 +149,15 @@ public static ProvisioningConfig buildConfig(ProvisioningManager pm,
                 String coords = getMavenCoords(fp);
                 fpl = FeaturePackLocation.fromString(coords);
             } else {
-                fpl = FeaturePackLocation.fromString(fp.getLocation());
+                //Special case for G:A that conflicts with producer:channel that we can't have in the plugin.
```
Comment (jmesnil) [inline core/src/main/java/org/wildfly/plugin/core/GalleonUtils.java:152]: @jfdenise this is a change for your PoC and I'm not sure if it's still valid. Could you please check?

File: examples/channel/pom.xml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        </dependency>
+    </dependencies>
+
+    <build>
+        <finalName>${project.artifactId}</finalName>
+        <plugins>
+            <plugin>
+                <groupId>org.wildfly.plugins</groupId>
+                <artifactId>wildfly-maven-plugin</artifactId>
+                <version>3.0.1-SNAPSHOT</version>
```
Comment (jmesnil) [inline examples/channel/pom.xml:36]: the examples POMs are independent of the maven-plugin-parent POM.
Updating the plugin version will make the examples out of sync. 
This needs to be resolved

File: core/src/main/java/org/wildfly/plugin/core/GalleonUtils.java (hunk 1 above)
Comment (jfdenise) [reply core/src/main/java/org/wildfly/plugin/core/GalleonUtils.java:152]: Yes, that is needed to not require to end the G:A with a ":".

**Comment by jfdenise** [comment]
@jmesnil @jamezp , it would be nice to bump WildFly maven plugin to let's say a 4.0.0.Alpha1 built with JDK11 containing this feature.

**Review by jfdenise — CHANGES_REQUESTED** [review]

File: examples/channel-example/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,2 @@
+FROM quay.io/wildfly/wildfly-runtime-jdk11
+COPY --chown=jboss:root target/server $JBOSS_HOME
```
[the remark this answers is not in the document]
Comment (jmesnil) [reply examples/channel-example/Dockerfile:2]: fixed

File: examples/jaxrs/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,2 @@
+FROM quay.io/wildfly/wildfly-runtime-jdk11
+COPY --chown=jboss:root target/server $JBOSS_HOME
```
[the remark this answers is not in the document]
Comment (jmesnil) [reply examples/jaxrs/Dockerfile]: fixed

File: plugin/src/main/java/org/wildfly/plugin/provision/ChannelMavenArtifactRepositoryManager.java [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+import org.jboss.galleon.universe.maven.MavenUniverseException;
+import org.jboss.galleon.universe.maven.repo.MavenRepoManager;
+import org.wildfly.channel.Channel;
+import org.wildfly.channel.ChannelMapper;
+import org.wildfly.channel.ChannelSession;
+import org.wildfly.channel.UnresolvedMavenArtifactException;
+import org.wildfly.channel.spi.MavenVersionsResolver;
+import org.wildfly.channel.version.VersionMatcher;
+
+public class ChannelMavenArtifactRepositoryManager implements MavenRepoManager {
```
[the remark this answers is not in the document]
Comment (jmesnil) [reply plugin/src/main/java/org/wildfly/plugin/provision/ChannelMavenArtifactRepositoryManager.java:63]: Good suggestion, tracked by https://github.com/wildfly-extras/wildfly-channel/issues/36

[the remark this answers is not in the document]
Comment (jmesnil) [reply plugin/src/main/java/org/wildfly/plugin/provision/ChannelMavenArtifactRepositoryManager.java:63]: > order for Galleon-plugins installer to detect that the channel are in place

I don't get that part. "Channels are in place" when their are in the plugin `<configuration>`. That's when we decide if we want to resolve artifacts using channels or doing it directly from the feature-pack's metadata. What would this marker interface change with that regard?

[the remark this answers is not in the document]
Comment (jfdenise) [reply plugin/src/main/java/org/wildfly/plugin/provision/ChannelMavenArtifactRepositoryManager.java:63]: The Galleon plugins: https://github.com/wildfly/galleon-plugins/tree/main/galleon-plugins, during installation, can take decision according to the fact that the feature-pack is provisioned with channels or without. An example is the expression resolution used for versions (datasources feature-pack). The expression must be resolved to a version when channels are not configued. When channels are configured, the expression is to be ignored (no need for system property or env variable to be set to define the version to provision), the channel rules all versions. Obviously if no stream is defined in a configured channel, the expression must be properly resolved.

**Comment by jmesnil** [comment]
@jfdenise I've released https://github.com/wildfly-extras/wildfly-channel/releases/tag/1.0.0.Alpha13 that incorporates your changes. I've updated this PR with this release (including the `org.wildfly.channel.spi.ChannelResolvable` marker interface). Could you please give it a second look?

**Comment by jamezp** [comment]
Overall this looks fine to me. The only thing I'm not sure about is having the example in the same project. I've always preferred to have examples separate from the project source. I don't have a real strong opinion on it though.

**Comment by jmesnil** [comment]
@jamezp the main argument against having them in their repo is that the examples are not tied to the parent POM and they can get out of sync if we do not update the plugin version in their pom.xml. Having them outside of the project might make more sense but where could we host them? Do you mean creating  new repo `wildfly/wildfly-maven-plugin-examples?

**Comment by jamezp** [comment]
> @jamezp the main argument against having them in their repo is that the examples are not tied to the parent POM and they can get out of sync if we do not update the plugin version in their pom.xml. Having them outside of the project might make more sense but where could we host them? Do you mean creating new repo `wildfly/wildfly-maven-plugin-examples`?

Hi @jmesnil. One thing I personally do no like is having standalone projects/modules depend on a SNAPSHOT artifact. It just feels not very user friendly. However, I realize that's a perfect world scenario which we do not always live in :)

As for where it could live outside of the project. I would be fine with a new repository or using the quickstarts if we feel that is the appropriate place.

**Comment by jmesnil** [comment]
@jamezp As discussed, I've removed the examples from the plugin source code. The quickstarts will eventually be updated to use the plugin to deploy WildFly so that's where the users will actually see the usage of the plugin.

**Review by jamezp — APPROVED** [review]

**Pull request merged**